File: py/ventora_billing/src/ventora_billing/trial.py
from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import UTC, datetime
from math import ceil
from typing import Protocol

logger = logging.getLogger(__name__)


@dataclass
class TrialRecord:
    user_id: str
    trial_ends_at: datetime
    status: str
    organization_id: str | None = None


class TrialDb(Protocol):
    def find_expired_trials(self, now: datetime) -> list[TrialRecord]: ...
    def mark_trial_expired(self, user_id: str) -> None: ...
    def find_trials_ending_soon(self, days_ahead: int, now: datetime) -> list[TrialRecord]: ...


class TrialEmailClient(Protocol):
    def send_trial_ending_warning(self, user_id: str, days_left: int) -> None: ...
    def send_trial_expired_notice(self, user_id: str) -> None: ...


@dataclass
class TrialLifecycleConfig:
    db: TrialDb
    email_client: TrialEmailClient | None = None
    warning_days_ahead: int = 3


class TrialLifecycle:
    def __init__(self, config: TrialLifecycleConfig) -> None:
        self._config = config

    def sweep_expired_trials(self, now: datetime | None = None) -> dict[str, int]:
        now = now or datetime.now(UTC)
        records = self._config.db.find_expired_trials(now)
        swept = 0
        for record in records:
            try:
                self._config.db.mark_trial_expired(record.user_id)
                swept += 1
            except Exception as e:
                logger.warning("Failed to expire trial for user %s: %s", record.user_id, e)
        return {"swept": swept}

    def dispatch_trial_reminders(self, now: datetime | None = None) -> dict[str, int]:
        if not self._config.email_client:
            return {"sent": 0}
        now = now or datetime.now(UTC)
        records = self._config.db.find_trials_ending_soon(self._config.warning_days_ahead, now)
        sent = 0
        for record in records:
            try:
                seconds_left = (record.trial_ends_at - now).total_seconds()
                days_left = max(0, ceil(seconds_left / 86_400))
                self._config.email_client.send_trial_ending_warning(record.user_id, days_left)
                sent += 1
            except Exception as e:
                logger.warning("Failed to send trial reminder for user %s: %s", record.user_id, e)
        return {"sent": sent}


def create_trial_lifecycle(config: TrialLifecycleConfig) -> TrialLifecycle:
    return TrialLifecycle(config)

File: packages/billing/src/trial.ts
// Trial lifecycle management — requires a database interface
// The DB interface is intentionally minimal (duck-typed) to avoid tying to Drizzle

export type TrialRecord = {
  userId: string;
  organizationId?: string;
  trialEndsAt: Date;
  status: "trialing" | "converted" | "expired";
};

export type TrialDb = {
  // Returns records where trialEndsAt < now and status === "trialing"
  findExpiredTrials(now: Date): Promise<TrialRecord[]>;
  // Update trial status to "expired"
  markTrialExpired(userId: string): Promise<void>;
  // Returns records where trialEndsAt is within warningDays days
  findTrialsEndingSoon(daysAhead: number, now: Date): Promise<TrialRecord[]>;
};

export type TrialEmailClient = {
  sendTrialEndingWarning(userId: string, daysLeft: number): Promise<void>;
  sendTrialExpiredNotice(userId: string): Promise<void>;
};

export type TrialLifecycleOpts = {
  db: TrialDb;
  emailClient?: TrialEmailClient;
  warningDaysAhead?: number; // default: 3
};

export interface TrialLifecycle {
  sweepExpiredTrials(now?: Date): Promise<{ swept: number }>;
  dispatchTrialReminders(now?: Date): Promise<{ sent: number }>;
}

export function createTrialLifecycle(opts: TrialLifecycleOpts): TrialLifecycle {
  const { db, emailClient, warningDaysAhead = 3 } = opts;

  return {
    async sweepExpiredTrials(now: Date = new Date()): Promise<{ swept: number }> {
      const expired = await db.findExpiredTrials(now);
      let swept = 0;
      for (const record of expired) {
        try {
          await db.markTrialExpired(record.userId);
          swept++;
        } catch (err) {
          console.warn(`Failed to mark trial expired for user ${record.userId}:`, err);
        }
      }
      return { swept };
    },

    async dispatchTrialReminders(now: Date = new Date()): Promise<{ sent: number }> {
      if (emailClient === undefined) return { sent: 0 };

      const ending = await db.findTrialsEndingSoon(warningDaysAhead, now);
      let sent = 0;
      for (const record of ending) {
        try {
          const msLeft = record.trialEndsAt.getTime() - now.getTime();
          const daysLeft = Math.max(0, Math.ceil(msLeft / (1000 * 60 * 60 * 24)));
          await emailClient.sendTrialEndingWarning(record.userId, daysLeft);
          sent++;
        } catch (err) {
          console.warn(`Failed to send trial reminder to user ${record.userId}:`, err);
        }
      }
      return { sent };
    },
  };
}
